### ISSUE TITLE
Decorators in AbstractScraper class not inherited by subclasses causing decorator code to be bypassed

### ISSUE THREAD

**Issue by aupmanyu** [hhursev/recipe-scrapers#355]
Hello,

Hope you are doing well. This is the first time I am raising an issue so apologies if I miss something.

A few salient points upfront: I am using 12.2.0 of the package, Python 3.7.2 with pipenv (version 2018.11.2) and Pycharm CE 2019.3.5.

I have recently started using this library and wanted to extend the `SchemaOrg` parser to also include other data. While investigating how to do this, I came across the below issue. 

I want to check if my understanding below is correct or if I have missed something obvious. Thanks. 

### ISSUE

> The `@Decorators.schema_org_priority` decorator on each method in Abstract class `AbstractScraper` is not called for subclasses (I tried primarily with `TheKitchn` class).

Having a quick look through the code of a few other classes, I feel that this issue may extend to other subclasses as well.

### POTENTIAL REASON
Based on my understanding, this is because the subclasses override parent methods and normally in Python3, subclasses do not inherit decorators since by the time the subclass is being instantiated, the decorator of the parent class has already executed and "replaced" the parent method with the decorated method. 

### POTENTIAL RESOLUTION
The way to have subclasses inherit decorators is using metaclasses (e.g. [here](https://stackoverflow.com/questions/18858759/python-decorating-a-class-method-that-is-intended-to-be-overwritten-when-inheri/18859019#18859019) and [here](https://stackoverflow.com/questions/57104276/python-subclass-method-to-inherit-decorator-from-superclass-method)).

The `AbstractScraper` already uses a metaclass (`ExceptionHandlingMetaclass`), which can be extended to also add `@Decorators.schema_org_priority` to the subclass methods (as Python3 does not readily support multiple metaclass inheritances).

### RESULTS WITH RESOLUTION
I have tried this and it gives the desired behaviour outlined in `Decorators` class i.e.:

- First the decorator function in `@Decorators.schema_org_priority` is run that tries to get data from `SchemaOrg`
- If this fails, it runs the code defined in the method in the subclass (i.e. directly parsing the soup)
- If the subclass doesn't have a method defined, it runs the method from the AbstractScraper class (relevant only for `links` and `site_name` for now).

Example code hack (`_exception_handling.py`):

```    
from ._decorators import Decorators
 
    def __new__(cls, class_name, bases, attributes):
        """
        Go through all class attributes.
        If the attribute name is in the ON_EXCEPTION_RETURN_VALUES.keys
        Decorate the "code" behind that attribute name with exception_handling.
        """
        for key, value in attributes.items():
            if key in ON_EXCEPTION_RETURN_VALUES.keys():
                attributes.update({key: exception_handling(Decorators.schema_org_priority(value))})  # ------- line added ------- #
                # attributes.update({key: exception_handling(value)})  # ------- line removed ------- #
        return type.__new__(cls, class_name, bases, attributes)
```

**Comment by hhursev** [comment]
You outline major things I'm not happy with too 😉  recipe-scrapers v13 will be major bump in functionality while keeping the old API intact.

Changes are on their way adding things called "plugins", allowing users to easily attach functionality of their own. By using pre/post scraping hooks for example. This will allow you to extend easily in future.

I'll ping you once the functionality is ready.

**Comment by aupmanyu** [comment]
Hi,

Thanks for getting back and sorry for the delay - that makes sense.

Let me know if you need a hand getting it live.

**Comment by jayaddison** [comment]
Closing a few old issues - please comment, re-open or file a new issue for any unresolved items (preferably after checking for more recent discussion!). Thank you.

(in this case: @aupmanyu - hopefully the plugin system mentioned here will help.  please let us know of any other problems you find)
